File: apps/api/src/routes/reconciliation.ts
import type { FastifyInstance } from 'fastify'
import type { AppContext } from '../context.js'
import {
  startMigrationSchema,
  transitionMigrationSchema,
  createNodeTransferSchema,
  transitionNodeTransferSchema,
  startReconciliationSchema,
  replayCheckpointSchema,
  createRecoverySchema,
} from '@atc/operations'
import { randomUUID } from 'crypto'

export function reconciliationRoutes(fastify: FastifyInstance, { ctx }: { ctx: AppContext }) {
  // ── Runtime Migrations ──────────────────────────────────────────────────────

  fastify.post('/api/v1/reconciliation/migrations/start', async (req, reply) => {
    if (!ctx.runtimeMigrationService) return reply.code(503).send({ error: 'Service unavailable' })
    const parsed = startMigrationSchema.safeParse(req.body)
    if (!parsed.success) return reply.code(400).send({ error: parsed.error.flatten() })
    const migration = await ctx.runtimeMigrationService.startMigration(parsed.data)
    return reply.code(201).send(migration)
  })

  fastify.post('/api/v1/reconciliation/migrations/:migrationId/complete', async (req, reply) => {
    if (!ctx.runtimeMigrationService) return reply.code(503).send({ error: 'Service unavailable' })
    const { migrationId } = req.params as { migrationId: string }
    const migration = await ctx.runtimeMigrationService.completeMigration(migrationId)
    return reply.code(200).send(migration)
  })

  fastify.post('/api/v1/reconciliation/migrations/:migrationId/fail', async (req, reply) => {
    if (!ctx.runtimeMigrationService) return reply.code(503).send({ error: 'Service unavailable' })
    const { migrationId } = req.params as { migrationId: string }
    const parsed = transitionMigrationSchema.safeParse({ migrationId, ...req.body as object })
    if (!parsed.success) return reply.code(400).send({ error: parsed.error.flatten() })
    const migration = await ctx.runtimeMigrationService.failMigration(migrationId, parsed.data.reason)
    return reply.code(200).send(migration)
  })

  fastify.get('/api/v1/reconciliation/migrations/:migrationId', async (req, reply) => {
    if (!ctx.runtimeMigrationService) return reply.code(503).send({ error: 'Service unavailable' })
    const { migrationId } = req.params as { migrationId: string }
    const migration = await ctx.runtimeMigrationService.getMigration(migrationId)
    if (!migration) return reply.code(404).send({ error: 'Not found' })
    return reply.code(200).send(migration)
  })

  // ── Node Transfers ──────────────────────────────────────────────────────────

  fastify.post('/api/v1/reconciliation/transfers/initiate', async (req, reply) => {
    if (!ctx.ownershipTransferService) return reply.code(503).send({ error: 'Service unavailable' })
    const parsed = createNodeTransferSchema.safeParse(req.body)
    if (!parsed.success) return reply.code(400).send({ error: parsed.error.flatten() })
    const transfer = await ctx.ownershipTransferService.initiateTransfer(parsed.data)
    return reply.code(201).send(transfer)
  })

  fastify.post('/api/v1/reconciliation/transfers/:transferId/complete', async (req, reply) => {
    if (!ctx.ownershipTransferService) return reply.code(503).send({ error: 'Service unavailable' })
    const { transferId } = req.params as { transferId: string }
    const transfer = await ctx.ownershipTransferService.completeTransfer(transferId)
    return reply.code(200).send(transfer)
  })

  fastify.post('/api/v1/reconciliation/transfers/:transferId/fail', async (req, reply) => {
    if (!ctx.ownershipTransferService) return reply.code(503).send({ error: 'Service unavailable' })
    const { transferId } = req.params as { transferId: string }
    const parsed = transitionNodeTransferSchema.safeParse({ transferId, status: 'failed' })
    if (!parsed.success) return reply.code(400).send({ error: parsed.error.flatten() })
    const transfer = await ctx.ownershipTransferService.failTransfer(transferId)
    return reply.code(200).send(transfer)
  })

  // ── Reconciliation Runtime ──────────────────────────────────────────────────

  fastify.post('/api/v1/reconciliation/run', async (req, reply) => {
    if (!ctx.crossNodeReconciliationService) return reply.code(503).send({ error: 'Service unavailable' })
    const parsed = startReconciliationSchema.safeParse(req.body)
    if (!parsed.success) return reply.code(400).send({ error: parsed.error.flatten() })
    const reconciliationId = parsed.data.reconciliationId ?? randomUUID()
    const result = await ctx.crossNodeReconciliationService.reconcile(
      reconciliationId,
      parsed.data.reconciliationType,
      parsed.data.regionId,
      parsed.data.serverId,
    )
    return reply.code(200).send(result)
  })

  // ── Snapshot Replay ─────────────────────────────────────────────────────────

  fastify.post('/api/v1/reconciliation/snapshots/replay', async (req, reply) => {
    if (!ctx.snapshotReplayService) return reply.code(503).send({ error: 'Service unavailable' })
    const parsed = replayCheckpointSchema.safeParse(req.body)
    if (!parsed.success) return reply.code(400).send({ error: parsed.error.flatten() })
    const replay = await ctx.snapshotReplayService.replayCheckpoint(
      parsed.data.entityId,
      parsed.data.snapshotId,
    )
    return reply.code(200).send(replay)
  })

  fastify.get('/api/v1/reconciliation/snapshots/pending', async (_req, reply) => {
    if (!ctx.snapshotReplayService) return reply.code(503).send({ error: 'Service unavailable' })
    const replays = await ctx.snapshotReplayService.listPendingReplays()
    return reply.code(200).send(replays)
  })

  // ── Runtime Recovery ────────────────────────────────────────────────────────

  fastify.post('/api/v1/reconciliation/recovery/start', async (req, reply) => {
    if (!ctx.runtimeRecoveryService) return reply.code(503).send({ error: 'Service unavailable' })
    const parsed = createRecoverySchema.safeParse(req.body)
    if (!parsed.success) return reply.code(400).send({ error: parsed.error.flatten() })
    const recovery = await ctx.runtimeRecoveryService.startRecovery(parsed.data)
    return reply.code(201).send(recovery)
  })

  // ── Runtime Consistency ─────────────────────────────────────────────────────

  fastify.get('/api/v1/reconciliation/consistency/check', async (_req, reply) => {
    if (!ctx.runtimeConsistencyService) return reply.code(503).send({ error: 'Service unavailable' })
    const result = await ctx.runtimeConsistencyService.validateConsistency()
    return reply.code(200).send(result)
  })
}

File: game/atc-core/server/reconciliation.lua
-- ATC Cross-Node Migration & Runtime Reconciliation Bridge
-- Phase 51

ATC.Reconciliation = {}

-- ── Runtime Migrations ───────────────────────────────────────────────────────

function ATC.Reconciliation.StartMigration(migrationNonce, entityId, fromServerId, toServerId, migrationData)
  return ATC.SDK.HTTP.Post('/api/v1/reconciliation/migrations/start', {
    migrationNonce = migrationNonce,
    entityId       = entityId,
    fromServerId   = fromServerId,
    toServerId     = toServerId,
    migrationData  = migrationData or {},
  })
end

function ATC.Reconciliation.CompleteMigration(migrationId)
  return ATC.SDK.HTTP.Post('/api/v1/reconciliation/migrations/' .. migrationId .. '/complete', {})
end

function ATC.Reconciliation.FailMigration(migrationId, reason)
  return ATC.SDK.HTTP.Post('/api/v1/reconciliation/migrations/' .. migrationId .. '/fail', {
    reason = reason,
  })
end

function ATC.Reconciliation.GetMigration(migrationId)
  return ATC.SDK.HTTP.Get('/api/v1/reconciliation/migrations/' .. migrationId)
end

-- ── Node Transfers ───────────────────────────────────────────────────────────

function ATC.Reconciliation.InitiateTransfer(entityId, fromServerId, toServerId, transferData)
  return ATC.SDK.HTTP.Post('/api/v1/reconciliation/transfers/initiate', {
    entityId     = entityId,
    fromServerId = fromServerId,
    toServerId   = toServerId,
    transferData = transferData or {},
  })
end

function ATC.Reconciliation.CompleteTransfer(transferId)
  return ATC.SDK.HTTP.Post('/api/v1/reconciliation/transfers/' .. transferId .. '/complete', {})
end

function ATC.Reconciliation.FailTransfer(transferId)
  return ATC.SDK.HTTP.Post('/api/v1/reconciliation/transfers/' .. transferId .. '/fail', {})
end

-- ── Reconciliation ───────────────────────────────────────────────────────────

function ATC.Reconciliation.Run(reconciliationType, regionId, serverId, reconciliationId)
  return ATC.SDK.HTTP.Post('/api/v1/reconciliation/run', {
    reconciliationType = reconciliationType,
    regionId           = regionId,
    serverId           = serverId,
    reconciliationId   = reconciliationId,
  })
end

-- ── Snapshot Replay ──────────────────────────────────────────────────────────

function ATC.Reconciliation.ReplayCheckpoint(entityId, snapshotId)
  return ATC.SDK.HTTP.Post('/api/v1/reconciliation/snapshots/replay', {
    entityId   = entityId,
    snapshotId = snapshotId,
  })
end

function ATC.Reconciliation.ListPendingReplays()
  return ATC.SDK.HTTP.Get('/api/v1/reconciliation/snapshots/pending')
end

-- ── Runtime Recovery ─────────────────────────────────────────────────────────

function ATC.Reconciliation.StartRecovery(entityId, recoveryType, targetServerId)
  return ATC.SDK.HTTP.Post('/api/v1/reconciliation/recovery/start', {
    entityId       = entityId,
    recoveryType   = recoveryType,
    targetServerId = targetServerId,
  })
end

-- ── Consistency Check ────────────────────────────────────────────────────────

function ATC.Reconciliation.CheckConsistency()
  return ATC.SDK.HTTP.Get('/api/v1/reconciliation/consistency/check')
end
